src/Rules.Framework/RulesEngine.cs

```diff
@@ -377,37 +410,92 @@ await ManagementOperations.Manage(existentRules)
                 .ConfigureAwait(false);
         }
 
-        private async Task<IEnumerable<Rule<TContentType, TConditionType>>> MatchAsync(
-            TContentType contentType,
-            DateTime matchDateBegin,
-            DateTime matchDateEnd,
-            IEnumerable<Condition<TConditionType>> conditions,
+        private IEnumerable<Rule<TContentType, TConditionType>> EvalAll(
+            List<Rule<TContentType, TConditionType>> orderedRules,
             EvaluationOptions evaluationOptions,
+            Dictionary<TConditionType, object> conditionsAsDictionary,
             bool? active)
         {
-            var getRulesArgs = new GetRulesArgs<TContentType>
+            // Begins evaluation at the first element of the given list as parameter. Returns all
+            // rules that match. Assumes given list is ordered.
+            var matchedRules = new List<Rule<TContentType, TConditionType>>(orderedRules.Count);
+            foreach (var rule in orderedRules)
             {
-                ContentType = contentType,
-                DateBegin = matchDateBegin,
-                DateEnd = matchDateEnd,
-            };
+                if (rule.Active == active.GetValueOrDefault(defaultValue: true) && (rule.RootCondition == null || this.conditionsEvalEngine.Eval(rule.RootCondition, conditionsAsDictionary, evaluationOptions)))
+                {
+                    matchedRules.Add(rule);
+                }
+            }
 
-            var rules = await this.rulesSource.GetRulesAsync(getRulesArgs).ConfigureAwait(false);
+            return matchedRules.AsReadOnly();
+        }
 
-            var conditionsAsDictionary = conditions.ToDictionary(ks => ks.Type, ks => ks.Value);
+        private Rule<TContentType, TConditionType> EvalOneReverse(
+            List<Rule<TContentType, TConditionType>> rules,
+            EvaluationOptions evaluationOptions,
+            Dictionary<TConditionType, object> conditionsAsDictionary,
+            bool? active)
+        {
+            // Begins evaluation at the last element of the given list as parameter. Returns the
+            // first rule that matches. Assumes given list is ordered.
+            for (int i = rules.Count - 1; i >= 0; i--)
+            {
+                var rule = rules[i];
+                if (rule.Active == active.GetValueOrDefault(defaultValue: true) && (rule.RootCondition == null || this.conditionsEvalEngine.Eval(rule.RootCondition, conditionsAsDictionary, evaluationOptions)))
+                {
+                    return rule;
+                }
```

Comment:
this content inside the for loop seems to be repeating on line 463. Is there a way to not duplicate the code?

Reply:
Unfortunately, it has to be duplicated, because the for loops have different conditions, and the bodies have a `return` instruction (which can't be extracted to a common method without sacrificing readability and performance).

I have, although, extracted the if condition to a method `EvalRule(...)`, so that this logic is not repeated on several locations.